cmd/kubeapps-apis/plugins/kapp_controller/packages/v1alpha1/server_ctrl_packages.go

```diff
@@ -509,3 +511,96 @@ func (s *Server) CreateInstalledPackage(ctx context.Context, request *corev1.Cre
 		InstalledPackageRef: installedRef,
 	}, nil
 }
+
+// UpdateInstalledPackage Updates an installed package managed by the 'kapp_controller' plugin
+func (s *Server) UpdateInstalledPackage(ctx context.Context, request *corev1.UpdateInstalledPackageRequest) (*corev1.UpdateInstalledPackageResponse, error) {
+	log.Infof("+kapp-controller UpdateInstalledPackage")
+
+	// Validate the request
+	if request == nil || request.GetInstalledPackageRef() == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "no request AvailablePackageRef provided")
+	}
+	if request.GetInstalledPackageRef().GetContext().GetNamespace() == "" || request.GetInstalledPackageRef().GetIdentifier() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "required context or identifier not provided")
+	}
+	if request.GetInstalledPackageRef().GetIdentifier() == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "no request Name provided")
+	}
+
+	packageRef := request.GetInstalledPackageRef()
+	cluster := packageRef.GetContext().GetCluster()
+	reconciliationOptions := request.GetReconciliationOptions()
+	pkgVersion := request.GetPkgVersionReference().GetVersion()
+	installedPackageName := request.GetInstalledPackageRef().GetIdentifier()
+	values := request.GetValues()
+	targetNamespace := request.GetInstalledPackageRef().GetContext().GetNamespace()
+
+	if cluster == "" {
+		cluster = s.globalPackagingCluster
+	}
+
+	typedClient, _, err := s.GetClients(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	// fetch the package install
+	pkgInstall, err := s.getPkgInstall(ctx, cluster, targetNamespace, installedPackageName)
+	if err != nil {
+		return nil, errorByStatus("get", "PackageInstall", installedPackageName, err)
+	}
+
+	// Update the values.yaml values file if any is passed, otherwise, delete the values
+	if values != "" {
+		secret, err := s.newSecret(installedPackageName, values, targetNamespace)
+		if err != nil {
+			return nil, errorByStatus("upsate", "Secret", secret.Name, err)
+		}
+		updatedSecret, err := typedClient.CoreV1().Secrets(targetNamespace).Update(ctx, secret, metav1.UpdateOptions{})
```

Comment:
Similar to the previous PR, I'd consider only updating the secret once the pkg install itself has successfully updated (if it is updating). We want to minimize the possibility of leaving an inconsistent state (and without transactions, we can't guarantee it).

Reply:
Yes, you're right. Moving the secret handling logic right after we perform the `PackageInstall` update.

I don't think we have transactions in k8s, but we might want to set some finalizers over the resources we create, so that everything got cleaned up automagically? This leads me to think... shouldn't be kapp-ctrl already handling it? should we add some special labels/annotations?

Reply:
Heh, yes, no transactions here.

> This leads me to think... shouldn't be kapp-ctrl already handling it? should we add some special labels/annotations?

I don't think we should add any kapp-ctrler annotations or such, but yep, perhaps (later) we can ensure that we know when a secret was created by Kubeapps on behalf of the user. Nothing for now though.